cli/lib/reload-cluster.js

```diff
@@ -1,393 +1,830 @@
+      this.items = []
+    }
+  }
+}
+
+
+
+
+class WorkerInfo {
+  //
+  constructor(worker) {
+    this.request_disconnect = false
+    this.request_shutdown = false
+    this.connectedEvent = worker.isConnected()
+    this.ready = false
+    this.trackingStartTime = Date.now()
+    this.worker_key = worker.id
+    this.address = ''
+  }
+  // --connectTimeout--------------------------------------- 
+  // Check if more time has gone by for connecting than we can tolerate 
+  connectTimeout() {
+    var worker = cluster.workers[this.worker_key]
+    if ( worker !== undefined ) {
+      if ( !(worker.isConnected()) ) {
+        var onset = this.trackingStartTime
+        var diff = Date.now() - onset
+        //
+        //if ( diff > MAX_CONNECT_FAIL_TIME  )  console.log(`DISCONNECT TIMEOUT ${this.worker_key}`)
+        //
+        return ( diff > MAX_CONNECT_FAIL_TIME ) 
       }
     } else {
-      process.nextTick(forceKillWorker);
+      return true
     }
+    return false
+  }
+  //
+}
+
+
+// GLOBAL
+// ---- ---- ---- ---- ---- ---- ----
+var gExtantTimers = new TimerList();
+var gStoppedProcCallbacks = new CallbackList();
+
+// keeping these as global variables.
+var tClosers = {}
+var tTracked = {}
+// ---- ---- ---- ---- ---- ---- ----
+
+
+// --shouldTrackWorker--------------------------------------- 
+// Decide if a worker should be put into or kept in the tracking map.
+function shouldTrackWorker(w,isTracked) {
+  if ( w === undefined ) return(false)
+  var w_info = tTracked[w.id]
+  var alreadyTracked = (w_info !== undefined) 
+  var wrongTrackingState = isTracked ? !alreadyTracked : alreadyTracked
+  var itsDead = w.isDead()
+  var itLostConnection = isTracked && !(w.isConnected() && (w_info && w_info.connectedEvent))
+  var itTimedoutWaitingForConnection = isTracked && (w_info && w_info.connectTimeout())
+  var itsInClosers = tClosers.hasOwnProperty(w.id)
+  //
+  var somethingsWrong = (wrongTrackingState || itsDead || itLostConnection || itTimedoutWaitingForConnection || itsInClosers)
+  //
+  return(!somethingsWrong)
+}
+
 
-    removeWorkerFromActiveWorkers(worker);
+// --cleanUpTrackedProcess--------------------------------------- 
+// look through the tracked processes to see if anyone has died or disconnected
+// called by consonantProcesses
+function cleanUpTrackedProcess() {  // walk through tracked processes
+  var cw_map = cluster.workers;
+  for ( var wk in tTracked ) {    // remove any processes that are dead and make room for new ones
+    var worker = cw_map[wk]  // processes are in tTracked immedately after a fork and retain space until they fail
+    if ( worker === undefined ) {  // process is gone 
+      delete tTracked[wk]  // stop tracking this process // and put it nowhere... it is gone (delete tracking info)
+    } else if ( !shouldTrackWorker(worker,true) ) {
+      // moved tracking info
+      tClosers[wk] = tTracked[wk] // this should never overwrite ... but it could be checkd. Decided not to.
```

Comment:
I don't understand this comment. What does "could be checkd. Decided not to." really mean? What are the consequences of this? (same comment exists below as well)